common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -309,41 +331,86 @@ protected boolean removeTask(Runnable task) {
      * @return {@code true} if and only if at least one task was run
      */
     protected boolean runAllTasks() {
+        assert inEventLoop();
         boolean fetchedAll;
+        boolean ranAtLeastOne = false;
+
         do {
             fetchedAll = fetchFromScheduledTaskQueue();
-            Runnable task = pollTask();
-            if (task == null) {
-                return false;
+            if (runAllTasksFrom(taskQueue)) {
+                ranAtLeastOne = true;
             }
+        } while (!fetchedAll); // keep on processing until we fetched all scheduled tasks.
 
-            for (;;) {
-                try {
-                    task.run();
-                } catch (Throwable t) {
-                    logger.warn("A task raised an exception.", t);
-                }
+        if (ranAtLeastOne) {
+            lastExecutionTime = ScheduledFutureTask.nanoTime();
+        }
+        afterRunningAllTasks(false);
+        return ranAtLeastOne;
+    }
 
-                task = pollTask();
-                if (task == null) {
-                    break;
-                }
+    /**
+     * Runs all tasks from the passed {@code taskQueue}.
+     *
+     * @param taskQueue To poll and execute all tasks.
+     *
+     * @return {@code true} if atleast one task was executed.
+     */
+    protected boolean runAllTasksFrom(Queue<Runnable> taskQueue) {
+        boolean ranAtLeastOne = false;
+        for (;;) {
+            Runnable task = pollTaskFrom(taskQueue);
+            if (task == null) {
+                break;
             }
-        } while (!fetchedAll); // keep on processing until we fetched all scheduled tasks.
-
-        lastExecutionTime = ScheduledFutureTask.nanoTime();
-        return true;
+            ranAtLeastOne = true;
```

Comment:
consider unrolling this loop ... this variable is only meaningful on the first iteration.

```java
    protected boolean runAllTasksFrom(Queue<Runnable> taskQueue) {
        Runnable task = pollTaskFrom(taskQueue);
        if (task == null) {
            return false;
        }
        safeExecute(task);
        for (;;) {
            task = pollTaskFrom(taskQueue);
            if (task == null) {
                return true;
            }
            safeExecute(task);
        }
    }

    private static void safeExecute(Runnable task) {
        try {
            task.run();
        } catch (Throwable t) {
            logger.warn("A task raised an exception. Task: {}", task, t);
        }
    }
```